Add `ServiceAccount`, `AuthorizationPolicy` and `Sidecar` manifests for asm-acm-tutorial

mathieu-benoit opened the pull request.

NimJay reviewed (changes requested):
I tested out the manifests (on my own GKE cluster) using the bash commands you listed.
It was a smooth experience! Everything worked.
Thank you for making this so easy to test! 🙌
I enjoy testing your pull-requests — I also learn so much. :)

**Parent Folder**
Right now the parent folder is `/demos/`.
But the [README.md in `/demos/`](https://github.com/GoogleCloudPlatform/anthos-service-mesh-samples/tree/main/demos) states:
> ... for content that does not live on cloud.google.com.

Since these samples are for a cloud.google.com tutorial, let's move these 3 folders (`asm-installation-manifests`, `ingress-gateway-asm-manifests`, `online-boutique-asm-manifests`) into `/docs/`. The [README.md in `/docs/`](https://github.com/GoogleCloudPlatform/anthos-service-mesh-samples/tree/main/docs):
> ... code samples for cloud.google.com.

(I feel like the folder names `demos` and `docs` are misleading — but changing that is a big project.)

**Many Resources**

This pull-request introduces 74+ files (i.e., a lot of different resources).
Will the tutorial include a diagram?
I think a diagram will be helpful.

**Link to Tutorial**

Maybe, once we release the tutorial, we can include a `README.md` in each of the 3 folders with a link to that tutorial. For instance, similar to [this README.md](https://github.com/GoogleCloudPlatform/kubernetes-engine-samples/tree/main/quickstart) and [this](https://github.com/GoogleCloudPlatform/kubernetes-engine-samples/tree/main/guestbook#guestbook-application-example).

**Additional Comments**
I've left more comments on the code itself.

mathieu-benoit commented on demos/online-boutique-asm-manifests/base/all/virtualservice.yaml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: networking.istio.io/v1alpha3
```
Just couple of files were missing that header, I just added them in order to close this one ;)

mathieu-benoit commented on demos/online-boutique-asm-manifests/authorization-policies/for-ingress-gateway/kustomization.yaml:
```diff
@@ -0,0 +1,26 @@
+      path: /spec/rules/0/from/0/source/principals
+      value:
+        - cluster.local/ns/ONLINEBOUTIQUE_NAMESPACE/sa/loadgenerator
+        - cluster.local/ns/INGRESS_GATEWAY_NAMESPACE/sa/INGRESS_GATEWAY_NAME
```
Good question, actually the default value in the `base` files are already set to these values, so by default we don't need to `sed`/override these values. I show the `sed` and overlays approach for scenario where user can deploy the apps in another namespace (`default`, `staging`, etc.) or have different names of their ingress gateway (`istio-ingress`, `istio-ingressgateway`, etc.). In the context of the tutorial we will use the default values.
Hope that makes sense and answers your question?

NimJay replied:
Ohhh, wow, you thought of everything.
Thanks, Mathieu. I'm glad you included the `sed` commands in this PR's description for demonstration (and my understanding).

mathieu-benoit commented on demos/online-boutique-asm-manifests/authorization-policies/kustomization.yaml:
```diff
@@ -0,0 +1,29 @@
+- all
+- for-ingress-gateway
+- for-namespace
+#- for-memorystore
```
There is different `for-memorystore` overlays in `base`, `sidecars` and `authorizationpolicies` folders, indeed. These are for the scenario where users don't want to use the in-cluster `redis` container, they could easily apply these `for-memorystore` overlays to properly configure their OnlineBoutique setup.
This is out-of-scope for the tutorial, but it's a good scenario to test the `Kustomize` approach with folder structure, `resources` versus `components` files, etc.
I could document how to chain this in order to test it as part of this PR, or like you suggested, it will part of a `README.md` file we could add later too.
Does that make sense? Let me know if you need complementary information about that.

NimJay replied:
Yes, this makes sense.
The `for-memorystore` flavour is opt-in — the others are default.

mathieu-benoit commented:
Rough notes about the structure we discussed @NimJay 
- docs/
  - asm-install-manifests
  - ingress-gateway-manifests
  - online-boutique-manifests

Later, we will add the tutorial assets in:
- docs/
  - asm-asm-tutorial/
    - online-boutique/
    - root-sync/

mathieu-benoit commented:
I think I took into account all your feedback and comments, let me know otherwise @NimJay. Thanks again!

I think also we could merge this PR into main once ready and not wait for the snippets/docs tags. We could do this in a separate PR later. No rush, but I think it could make sense, especially since I will soon open another PR for the dedicated tutorial content.

NimJay reviewed (approved):
Thanks for performing the folder restructuring, and thanks for talking through my concerns during our last 1-on-1.
I approve. I think, given that this is one of the biggest pull-requests, we should have another DPE's review and approval before merging.

Shabirmean commented on docs/online-boutique-asm-manifests/base/kustomization.yaml:
```diff
@@ -0,0 +1,31 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Sample to illustrate how to customize the OnlineBoutique namespace and the ASM_VERSION sidecar injection, you need to replace:
+# ONLINEBOUTIQUE_NAMESPACE in for-namespace/kustomization.yaml
+# and ONLINEBOUTIQUE_NAMESPACE and ASM_VERSION in for-asm-channel/kustomization.yaml
+# before deploying this file.
+# As an example:
+# sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g" for-namespace/kustomization.yaml
+# sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g;s/ASM_VERSION/asm-managed-rapid/g" for-asm-channel/kustomization.yaml
```
Same as previous comments. I would suggest removing the examples with `sed` from inside the yaml files. I would do that for all the files. 

Also I feel the line `Sample to illustrate how to customize the OnlineBoutique namespace `` does not tell the user much.

mathieu-benoit replied:
Done.

Shabirmean commented on docs/online-boutique-asm-manifests/kustomization.yaml:
```diff
@@ -0,0 +1,44 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Sample to illustrate how to customize the OnlineBoutique deployments, you need to replace:
+# ONLINEBOUTIQUE_NAMESPACE in this current kustomization.yaml
+# ONLINEBOUTIQUE_NAMESPACE in base/for-namespace/kustomization.yaml
+# ONLINEBOUTIQUE_NAMESPACE and ASM_VERSION in base/for-asm-channel/kustomization.yaml
+# ONLINEBOUTIQUE_NAMESPACE in authorization-policies/for-namespace/kustomization.yaml
+# ONLINEBOUTIQUE_NAMESPACE, INGRESS_GATEWAY_NAMESPACE and INGRESS_GATEWAY_NAME in authorization-policies/for-ingress-gateway/kustomization.yaml
+# ONLINEBOUTIQUE_NAMESPACE in sidecars/for-namespace/kustomization.yaml
+# before deploying this file.
+# As an example:
+# sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g" kustomization.yaml
+# sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g" base/for-namespace/kustomization.yaml
+# sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g;s/ASM_VERSION/asm-managed-rapid/g" base/for-asm-channel/kustomization.yaml
+# sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g" authorization-policies/for-namespace/kustomization.yaml
+# sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g;s/INGRESS_GATEWAY_NAMESPACE/asm-ingress/g;s/INGRESS_GATEWAY_NAME/asm-ingressgateway/g" authorization-policies/for-ingress-gateway/kustomization.yaml
+# sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g" sidecars/for-namespace/kustomization.yaml
```
Same as before. Suggestion to remove the examples from inside the yaml files. Ideally I would put them on a README file at the root of every Kustomize file. 

Also having these here makes the string `ONLINEBOUTIQUE_NAMESPACE` get changed whenever the `sed` command is run to replace the strings

mathieu-benoit replied:
Done.

Shabirmean commented on docs/online-boutique-asm-manifests/sidecars/kustomization.yaml:
```diff
@@ -0,0 +1,23 @@
+components:
+- all
+- for-namespace
+#- for-memorystore
```
Do we need this? If so can we add a comment about it?

mathieu-benoit replied:
I removed it, and created a dedicated optional section about that `for-memorystore` setup.

Shabirmean commented on docs/online-boutique-asm-manifests/sidecars/kustomization.yaml:
```diff
@@ -0,0 +1,23 @@
+# limitations under the License.
+
+# Sample to illustrate how to customize the OnlineBoutique namespace, you need to replace the variable ONLINEBOUTIQUE_NAMESPACE in for-namespace/kustomization.yaml before deploying this file.
+# As an example: sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g" for-namespace/kustomization.yaml
```
Same comment as before on adding examples inside the yaml files

mathieu-benoit replied:
Done.

Shabirmean reviewed (changes requested):
Thank you very much for submitting this immense improvement @mathieu-benoit. I have left a few comments throughout the yaml files.

#### Summary:
- I was able to fully run through the example steps you had provided. The cluster seems to be healthy and everything was okay
- I saw some commented out lines inside the yaml definition. I was not sure if they were required lines. If so can we add some comments to explain what they are for?
- I also feel that we should remove example code/commands from inside the yaml files. Also the `sed` command changes the values in the comments as well, corrupting the example itself.
- When we publish this as a tutorial, is there a way we can use something else instead of `sed`, since the syntax for sed is different between MacOS and other linux distros



#### Key error to fix:
1. First time when you deploy the manifests with `kubectl apply -k asm-installation-manifests/` you get the following error. This is mainly because it takes a while for the ASM setup to complete. So I think we should separate this into two steps. If not it looks like something didn't work.
```sh
unable to recognize "asm-installation-manifests/": no matches for kind "AuthorizationPolicy" in version "security.istio.io/v1beta1"
unable to recognize "asm-installation-manifests/": no matches for kind "PeerAuthentication" in version "security.istio.io/v1beta1"
```

2. Can we add top level READMEs in some of the directories here to better explain what these yaml files are? I think that README will serve a better purpose than the actual commented explanation on top of each yaml file. I feel that we can explain the information about certain placeholder fields in the yamls needing to be removed also in the README. 

- `docs/asm-installation-manifests`: explain what the yaml files are, then what is the content of each of the sub-directories
- `docs/online-boutique-asm-manifests`: explain what the yaml files are, then what is the content of each of the sub-directories

mathieu-benoit commented on docs/asm-installation-manifests/kustomization.yaml:
```diff
@@ -0,0 +1,26 @@
+# limitations under the License.
+
+# Sample to illustrate how to customize the ASM channel, you need to replace the variables ASM_VERSION and ASM_CHANNEL in for-asm-channel/kustomization.yaml before deploying this file.
+# As an example: sed -i "s/ASM_VERSION/asm-managed-rapid/g;s/ASM_CHANNEL/rapid/g" for-asm-channel/kustomization.yaml
```
Good point about MacOS, so typically replacing `""` by `''`, right?

mathieu-benoit replied:
Done.

mathieu-benoit replied:
After some tests, `''` doesn't work on Linux/Cloud Shell if there is a bash variable in there.

After chatting about that with TWs, it appears that because we are targeting Cloud Shell, we are fine with keeping this `sed -i ""...` approach.

With that said, that's really good to know for MacOS use case, thanks for raising this.

NimJay replied:
I think Shabir might be talking about the need for an extra pair of quotations (as described in [this StackOverflow post](https://unix.stackexchange.com/questions/401905/bsd-sed-vs-gnu-sed-and-i/401928#401928)).

But no changes necessary since Nicola (TW) said we should be encouraging (and primarily supporting) Cloud Shell anyway.

We can resolve this thread IMO.

mathieu-benoit commented on docs/online-boutique-asm-manifests/base/for-memorystore/kustomization.yaml:
```diff
@@ -0,0 +1,44 @@
+      path: /spec/template/spec/containers/0/env/0
+      value:
+        name: REDIS_ADDR
+        value: REDIS_IP:REDIS_PORT
```
I didn't think about that, that's a good point. Now I think more about this, Memorystore has 2 flavors: Memcache and Redis and I'm not sure `cartservice` will work with Memcache (?). Furthermore, it could be any Redis database, hosted elsewhere, like in a VM, etc.  So I think keeping this `REDIS_` prefix generic still makes more sense, isn't it?

Note: actually the value of `REDIS_IP` could be an host/fqdn, so wondering if we want to rename `REDIS_IP` by `REDIS_HOST` to be more generic? Memorystore (redis) only supports IP address anyway... so maybe we keep this like this?

mathieu-benoit commented on docs/online-boutique-asm-manifests/base/for-memorystore/kustomization.yaml:
```diff
@@ -0,0 +1,44 @@
+    - op: replace
+      path: /spec/template/spec/containers/0/env/0
+      value:
+        name: REDIS_ADDR
```
`REDIS_ADDR` is the actual name of the environment variable in the `cartservice` `Deployment` manifest, upstream.

NimJay replied:
For convenience, some links: 
* link to [cartservice.yaml](https://github.com/GoogleCloudPlatform/microservices-demo/blob/release/v0.3.6/kubernetes-manifests/cartservice.yaml#L37) that sets `REDIS_ADDR` to `redis-cart:6379` for normal/vanilla/default deployments.

mathieu-benoit commented on docs/online-boutique-asm-manifests/base/all/kubernetes-manifests.yaml:
```diff
@@ -0,0 +1,749 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# ----------------------------------------------------------
```
Yeah, thought about that, but don't want to update it, as I would like to get it upstream as-is once a new upstream release is out.

mathieu-benoit replied:
Done, in `online-boutique-asm-manifests/README.md` file.

Shabirmean replied:
Thank you!

mathieu-benoit commented on docs/online-boutique-asm-manifests/base/kustomization.yaml:
```diff
@@ -0,0 +1,31 @@
+- for-asm-channel
+- for-namespace
+#- for-virtualservice-host
+#- for-memorystore
```
I removed them, and created dedicated optional sections about that `for-virtualservice-host` and `for-memorystore` setups.

mathieu-benoit commented on docs/online-boutique-asm-manifests/base/for-memorystore/kustomization.yaml:
```diff
@@ -0,0 +1,44 @@
+# Sample to illustrate how to customize the OnlineBoutique's cartservice to use Memorystore (redis) instead of in-cluster redis container, you need to replace:
+# REDIS_IP and REDIS_PORT in this current file before including this file in another kustomization.yaml file.
+# As an example:
+# sed -i "s/REDIS_IP/0.0.0.0/g;s/REDIS_PORT/6379/g" kustomization.yaml
```
Done.

mathieu-benoit commented on docs/online-boutique-asm-manifests/authorization-policies/kustomization.yaml:
```diff
@@ -0,0 +1,29 @@
+# and ONLINEBOUTIQUE_NAMESPACE, INGRESS_GATEWAY_NAMESPACE and INGRESS_GATEWAY_NAME in for-ingress-gateway/kustomization.yaml
+# before deploying this file.
+# As an example:
+# sed -i "s/ONLINEBOUTIQUE_NAMESPACE/onlineboutique/g" for-namespace/kustomization.yaml
```
Done.

mathieu-benoit commented on docs/online-boutique-asm-manifests/authorization-policies/kustomization.yaml:
```diff
@@ -0,0 +1,29 @@
+- all
+- for-ingress-gateway
+- for-namespace
+#- for-memorystore
```
DOne.

mathieu-benoit commented:
Thanks @Shabirmean for your comments, feedback and suggestions, very hepful! I tried to update that repo accordingly, do you mind give it a last review (without testing maybe)? Thanks!
Also @NimJay, let me know if there is anything else to take into account here, thanks again too!

NimJay commented:
Thanks for addressing all our comments, @mathieu-benoit!

@Shabirmean, thanks for your elaborate review! As Mathieu mentioned, feel free to skip the final manual test (i.e., deploying in your own cluster) — we will be doing more tests of these samples in the coming weeks as we write (and review) the accompanying cloud.google.com tutorials.

Shabirmean commented:
@mathieu-benoit - Thank you very much for patiently responding to all the comments and making the necessary updates. Very much appreciate it. All the changes look good. I have 2 comments. One of them, we might want to verify before merging and the other we can followup later.

1. **Check before merge:** I tried to redo the tutorial and it seems like the resource are not getting deployed in the `onlineboutique` namespace. Instead they are all in `default` When I try the command `+ kubectl wait --for=condition=available --timeout=600s deployment --all -n onlineboutique` I can an error:
```sh
error: no matching resources found
```

2. **Follow up later:** Thank you very much for updating the README's with some details about the artifacts. However, I feel to make them more consumable we would want to be a little bit detailed about the folder structure, what the intention is and how they combine together. In it's current state _(the README)_, I think there is an expectation to have a basic idea of what these are. However, we would want to cater to people landing at this repo without a lot of background to be able to grasp what these manifests are why they are used. This is a practice I think we can grow going forward when adding new artifacts to this repo. I have created a followup issues for it: https://github.com/GoogleCloudPlatform/anthos-service-mesh-samples/issues/44

mathieu-benoit commented:
Thanks @Shabirmean!

For 1., good catch, the set of the namespace was missing, I just updated the script for the recommended test accordingly. Furthermore, while re-doing the e2e test, I found out that a typo was made recently, that I just fixed with my last commit.

For 2,, totally makes sense, we could improve the documentation of this in this repo, like you suggested and like discussed, we will improve that part in a future PR.

Thanks again! Merging this PR into `main` now.

Shabirmean reviewed (approved):
LGTM!
